File: src/main/java/AoC2017_05.java
import java.util.Arrays;
import java.util.List;
import java.util.function.IntUnaryOperator;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2017_05 extends SolutionBase<int[], Integer, Integer> {

    private AoC2017_05(final boolean debug) {
        super(debug);
    }

    public static AoC2017_05 create() {
        return new AoC2017_05(false);
    }

    public static AoC2017_05 createDebug() {
        return new AoC2017_05(true);
    }
 
    @Override
    protected int[] parseInput(final List<String> inputs) {
        return inputs.stream().mapToInt(Integer::parseInt).toArray();
    }

    private int countJumps(
            final int[] input,
            final IntUnaryOperator jumpCalculator
    ) {
        final int[] offsets = Arrays.copyOf(input, input.length);
        int cnt = 0;
        int i = 0;
        while (i < offsets.length) {
            final int jump = offsets[i];
            offsets[i] = jumpCalculator.applyAsInt(jump);
            i += jump;
            cnt++;
        }
        return cnt;
    }
    
    @Override
    public Integer solvePart1(final int[] input) {
        return countJumps(input, jump -> jump + 1);
    }
    
    @Override
    public Integer solvePart2(final int[] input) {
        return countJumps(input, jump -> jump >= 3 ? jump - 1 : jump + 1);
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "5"),
        @Sample(method = "part2", input = TEST, expected = "10"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2017_05.create().run();
    }
    
    private static final String TEST = """
            0
            3
            0
            1
            -3
            """;
}

File: src/main/python/AoC2017_05.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 5
#

import sys
from typing import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = list[int]
Output1 = int
Output2 = int


TEST = """\
0
3
0
1
-3
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [int(_) for _ in input_data]

    def count_jumps(
        self, input: list[int], jump_calculator: Callable[[int], int]
    ) -> int:
        offsets = [_ for _ in input]
        i = 0
        cnt = 0
        while i < len(offsets):
            jump = offsets[i]
            offsets[i] = jump_calculator(jump)
            i += jump
            cnt += 1
        return cnt

    def part_1(self, input: Input) -> int:
        return self.count_jumps(input, lambda j: j + 1)

    def part_2(self, input: Input) -> int:
        return self.count_jumps(input, lambda j: j - 1 if j >= 3 else j + 1)

    @aoc_samples(
        (
            ("part_1", TEST, 5),
            ("part_2", TEST, 10),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2017, 5)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
